### PR TITLE
Transacation and Block tests

### PR DESCRIPTION
**Pull request by JRafaelSoares**
Issue #15

### CODE REVIEW

File: blockchain/blockchain_data_structure.py [hunk 1]
```diff
@@ -1,59 +1,101 @@
...
 
+    def check_arguments(self, from_address, to_address, amount, node_id):
+        if not to_address:
+            raise Exception("Transaction must have a destination address")
+
+        if not node_id:
+            raise Exception("Transaction must have an associated node id")
+
+        # What type should from and to address be? How do we define public keys?
+        # TODO - Add from and to address type checks
```
Comment (RafaelAPB) [inline blockchain/blockchain_data_structure.py:31]: create issue?

File: blockchain/blockchain_data_structure.py [hunk 2]
```diff
@@ -1,59 +1,101 @@
...
+
+        # What type we want amount to be? Probably not int
+        # TODO - Add amount type check
```
Comment (RafaelAPB) [inline blockchain/blockchain_data_structure.py]: create issue?
ammount could be float

File: blockchain/blockchain_data_structure.py [hunk 3]
```diff
@@ -1,59 +1,101 @@
...
             return False
 
         h = SHA256.new()
-        h.update("self.id + self.fromAddress + self.toAddress + str(self.amount)".encode())
-        verify_sig(h, self.signature, node_id)  # Throws error if signature is invalid
+        h.update(self.transaction_content().encode())
+
+        try:
+            verify_sig(h, self.signature, self.node_id)  # Throws error if signature is invalid
+        except Exception:
```
Comment (RafaelAPB) [inline blockchain/blockchain_data_structure.py:54]: Return more info

File: blockchain/blockchain_data_structure.py [hunk 4]
```diff
@@ -1,59 +1,101 @@
...
+
+        if index < 0 or not type(index) is int:
+            raise Exception("Index must be a positive integer greater or equal to 0")
+
+        try:
+            self.has_valid_transactions(transactions)
+        except:
+            raise Exception("Invalid Transactions")
+
+        # TODO - Add check for previous hash to be empty only for first block
```
Comment (RafaelAPB) [inline blockchain/blockchain_data_structure.py:94]: Those todos are excellent catches: perhaps add one issue if you are not planning to address them atm

**Review by RafaelAPB — COMMENTED** [review]
Fantastic work!

**Review by CatarinaPedreira — DISMISSED** [review]

**Comment by CatarinaPedreira** [comment]
LGTM ;)

File: blockchain/blockchain_data_structure.py (hunk 2 above)
Comment (JRafaelSoares) [reply blockchain/blockchain_data_structure.py]: Ammount changed to float and check added!

File: blockchain/blockchain_data_structure.py (hunk 3 above)
Comment (JRafaelSoares) [reply blockchain/blockchain_data_structure.py:54]: Added a better error information (although we cannot know why it actually fails, either by integrity of wrong ID)

File: blockchain/blockchain_data_structure.py (hunk 1 above)
Comment (JRafaelSoares) [reply blockchain/blockchain_data_structure.py:31]: Issue created

File: blockchain/blockchain_data_structure.py (hunk 4 above)
Comment (JRafaelSoares) [reply blockchain/blockchain_data_structure.py:94]: Issues added!

**Pull request merged**